### PR TITLE
ADR0002: document deprecation strategy for current release series post 1.0

### LINKED ISSUE #1127: Decide on deprecation strategy
The current plan is to develop the new TUF alongside the old TUF (see #1126). 
How long *after* the refactor is complete will we continue to support the old code base? What does support look like? Bug fixes? Security fixes? We need a roadmap there. The main reason to keep the old code base around is to have a client that supports Python 2.7 (see  #1125). Not sure if this is a good reason for very much longer.

Independently of the refactor, we might generally want to think of a deprecation strategy for the future. E.g. do we want to display deprecation warnings for *n* releases before we drop interface functions? This probably only makes sense if we follow a stricter release schedule than we do now.

### PR DESCRIPTION
**Pull request by joshuagl**
**Description of the changes being introduced by the pull request**:

Per the discussion in #1127 opt to support the old release on a best-effort
basis.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Review by trishankatdatadog — COMMENTED** [review]
Thanks for documenting this, Joshua!

Do we have a policy for backporting security fixes?

**Comment by joshuagl** [comment]
Do you think security fixes should be handled differently to what is proposed here? If so, what additional steps do you think the maintainers should commit to?

I'm absolutely open to input here, I'm just not sure what to propose we commit to.

**Comment by trishankatdatadog** [comment]
> Do you think security fixes should be handled differently to what is proposed here? If so, what additional steps do you think the maintainers should commit to?
> 
> I'm absolutely open to input here, I'm just not sure what to propose we commit to.

I think we should at least commit in writing to prioritizing and guaranteeing security fixes for Python 2, but nothing else. WDYT?

**Comment by joshuagl** [comment]
> > Do you think security fixes should be handled differently to what is proposed here? If so, what additional steps do you think the maintainers should commit to?
> > I'm absolutely open to input here, I'm just not sure what to propose we commit to.
> 
> I think we should at least commit in writing to prioritizing and guaranteeing security fixes for Python 2, but nothing else. WDYT?

Prioritising and guaranteeing review/merge/release of security fixes seems reasonable. Is that what you're suggesting? Or did you mean to also suggest implementation of security fixes?

**Comment by trishankatdatadog** [comment]
> Prioritising and guaranteeing review/merge/release of security fixes seems reasonable. Is that what you're suggesting? Or did you mean to also suggest implementation of security fixes?

Both

**Comment by mnm678** [comment]
I think that we can reasonable commit to review and merge security fixes, but actually implementing the fixes may require more resources. To mitigate this, we can adopt a reporting strategy so that any users of the Python 2 implementation will be notified before any security issues become public, and will have a change to submit a patch.

**Comment by trishankatdatadog** [comment]
> I think that we can reasonable commit to review and merge security fixes, but actually implementing the fixes may require more resources. To mitigate this, we can adopt a reporting strategy so that any users of the Python 2 implementation will be notified before any security issues become public, and will have a change to submit a patch.

If someone else privately reports a security issue to us, then we could ask them to submit a patch, too. But what if we find the issue ourselves?

**Comment by lukpueh** [comment]
Why the change of heart, @trishankatdatadog?

@joshuagl already described the strategy proposed here in https://github.com/theupdateframework/tuf/issues/1127#issuecomment-713663707, and you seemed to deem it reasonable. (https://github.com/theupdateframework/tuf/issues/1127#issuecomment-713670216).

I admit that in the original proposal the best-effort basis sounded a bit more optimistic ...
> * that we will review any functional bug and security fixes proposed to the branch (PRs)
> * that severe security issues and functional bugs will be addressed on a best effort basis
> * that we will release updates with any generated fixes in a timely fashion

... than it does now ...

>  Bugs reported with tuf versions prior to 1.0.0 will likely not be addressed directly by tuf’s maintainers.

But I think the essence is the same.

**Review by lukpueh — COMMENTED** [review]
Many thanks for the great write-up, @joshuagl! In #1127 you talked about:
> a deprecation strategy for the current release series [...] and
> a deprecation/API stability strategy for releases after the refactor.

IIUC this document only addresses the former. Would you mind clarifying the distinction in the title (and maybe document name)?

**Comment by trishankatdatadog** [comment]
> Why the change of heart, @trishankatdatadog?

Sorry, what change of heart?

**Comment by lukpueh** [comment]
> Sorry, what change of heart?

Oh, apologies, I missed an important pronoun in my comment above.

I was just curious why you now wanted to commit to more (https://github.com/theupdateframework/tuf/pull/1203#issuecomment-726801429) than what @joshuagl had already described in https://github.com/theupdateframework/tuf/issues/1127#issuecomment-713663707, which you seemed to deem reasonable (https://github.com/theupdateframework/tuf/issues/1127#issuecomment-713670216).

**Comment by trishankatdatadog** [comment]
> I was just curious why you now wanted to commit to more ([#1203 (comment)](https://github.com/theupdateframework/tuf/pull/1203#issuecomment-726801429)) than what @joshuagl had already described in [#1127 (comment)](https://github.com/theupdateframework/tuf/issues/1127#issuecomment-713663707), which you seemed to deem reasonable ([#1127 (comment)](https://github.com/theupdateframework/tuf/issues/1127#issuecomment-713670216)).

I see. Yes, I still agree. I just think it's important to clearly state we should make an exception for security fixes, just like [Python 2](https://www.python.org/doc/sunset-python-2/)... until I saw that, nope, they don't guarantee security fixes either. Huh, wow. If that's the case, it may not be a terrible precedent. May encourage people to seriously upgrade. Not a bad idea. I know we're a more security-based project, but that doesn't mean we have to support everything forever. Ok, let's stick to the original plan.

**Review by trishankatdatadog — APPROVED** [review]

File: docs/adr/0002-deprecation-strategy.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+* Date: 2020-11-05
+
+Technical Story: https://github.com/theupdateframework/tuf/issues/1127
+
+## Context and Problem Statement
+
+We plan to refactor the reference implementation significantly and, as part of
+that effort, drop support for no-longer maintained versions of Python
+(see ADR 0002).
```
Comment (sechkova) [inline docs/adr/0002-deprecation-strategy.md]: I think this should be a reference to ADR 0001

Comment (joshuagl) [reply docs/adr/0002-deprecation-strategy.md]: Absolutely right, good catch!

**Comment by joshuagl** [comment]
Thank you all for continuing the discussion in my absence.

> Many thanks for the great write-up, @joshuagl! In #1127 you talked about:
> 
> > a deprecation strategy for the current release series [...] and
> > a deprecation/API stability strategy for releases after the refactor.
> 
> IIUC this document only addresses the former. Would you mind clarifying the distinction in the title (and maybe document name)?

Well spotted, thank you for the thorough review. I've pushed a fixup commit 8685898 which renames the document and changes the document title.

File: docs/adr/index.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ This log lists the architectural decisions for tuf.
 
 - [ADR-0000](0000-use-markdown-architectural-decision-records.md) - Use Markdown Architectural Decision Records
 - [ADR-0001](0001-python-version-3-6-plus.md) - Default to Python 3.6 or newer for new development
+- [ADR-0002](0002-deprecation-strategy.md) - Deprecation strategy
```
Comment (lukpueh) [inline docs/adr/index.md]:
```suggestion
- [ADR-0002](0002-pre-1-0-deprecation-strategy.md) - Deprecation strategy
```

**Review by lukpueh — CHANGES_REQUESTED** [review]
Please update link, other than that LGTM! :)

**Review by lukpueh — APPROVED** [review]

**Comment by lukpueh** [comment]
Thanks for the updates, @joshuagl. No need to wait for appveyor here. Merging...

**Pull request merged**